File: javascript/src/transport.ts
import { AuthManager, type AuthSession } from './auth.js';
import { JustDeployError, JustDeployValidationError } from './errors.js';
import type { RequestOptions } from './types.js';
import { SDK_HEADER } from './version.js';

type Method = 'GET' | 'POST' | 'PUT' | 'DELETE';
const API_TIMEOUT_MS = 30_000;

interface ApiRequestOptions extends RequestOptions {
  body?: unknown;
  headers?: Readonly<Record<string, string>>;
}

async function responsePayload(response: Response): Promise<unknown> {
  const text = await response.text();
  if (text.length === 0) return undefined;
  try {
    return JSON.parse(text) as unknown;
  } catch {
    throw new JustDeployError('JustDeploy returned a response that was not valid JSON.', { status: response.status });
  }
}

function apiError(response: Response, payload: unknown): JustDeployError {
  const details = payload && typeof payload === 'object' && !Array.isArray(payload) ? (payload as Record<string, unknown>) : {};
  const message = typeof details.message === 'string' && details.message.length > 0 ? details.message : `JustDeploy request failed with status ${response.status}.`;
  return new JustDeployError(message, {
    status: response.status,
    retryAfter: typeof details.retryAfter === 'number' ? details.retryAfter : null,
    requestId: typeof details.requestId === 'string' ? details.requestId : (response.headers.get('x-request-id') ?? null),
    details,
  });
}

function assertInternalPath(path: string): void {
  if (!path.startsWith('/') || path.startsWith('//') || path.includes('://') || path.includes('\\')) {
    throw new Error('Invalid internal JustDeploy API path.');
  }
}

export class Transport {
  readonly auth: AuthManager;
  private readonly fetcher: typeof fetch;

  constructor(auth: AuthManager, fetcher: typeof fetch = globalThis.fetch) {
    this.auth = auth;
    this.fetcher = fetcher;
  }

  async organizationRequest<T>(method: Method, path: string, options: ApiRequestOptions = {}): Promise<T> {
    assertInternalPath(path);
    const session = await this.auth.getSession();
    return this.send<T>(method, path, session, options, false);
  }

  private async send<T>(method: Method, path: string, session: AuthSession, options: ApiRequestOptions, replayed: boolean): Promise<T> {
    const resolved = new URL(`/organizations/${encodeURIComponent(session.organizationId)}${path}`, session.apiOrigin);
    const headers: Record<string, string> = {
      accept: 'application/json',
      authorization: `Bearer ${session.token}`,
      'x-justdeploy-sdk': SDK_HEADER,
    };
    for (const [name, value] of Object.entries(options.headers ?? {})) {
      const normalized = name.toLowerCase();
      if (normalized === 'authorization' || normalized === 'host' || normalized === 'x-justdeploy-sdk') {
        throw new Error(`The internal header ${name} cannot be overridden.`);
      }
      headers[normalized] = value;
    }
    let body: string | undefined;
    try {
      body =
        options.body === undefined
          ? undefined
          : JSON.stringify(options.body, (_key, value: unknown) => {
              if (typeof value === 'number' && !Number.isFinite(value)) throw new Error('invalid number');
              return value;
            });
    } catch {
      throw new JustDeployValidationError('The request contains a value that cannot be encoded as JSON.');
    }
    if (body !== undefined) headers['content-type'] = 'application/json';

    const request: RequestInit = { method, headers, redirect: 'error' };
    if (body !== undefined) request.body = body;
    const timeout = AbortSignal.timeout(API_TIMEOUT_MS);
    const requestSignal = options.signal ? AbortSignal.any([options.signal, timeout]) : timeout;
    request.signal = requestSignal;

    let response: Response;
    try {
      response = await this.fetcher(resolved, request);
    } catch {
      const message = options.signal?.aborted
        ? 'The JustDeploy request was canceled.'
        : requestSignal.aborted
          ? 'The JustDeploy request timed out.'
          : 'The JustDeploy request failed before the server returned a response.';
      throw new JustDeployError(message);
    }

    if (response.status === 401 && method === 'GET' && !replayed) {
      await response.body?.cancel().catch(() => undefined);
      const refreshed = await this.auth.refreshAfterUnauthorized(session.token);
      return this.send<T>(method, path, refreshed, options, true);
    }

    const payload = await responsePayload(response);
    if (!response.ok) throw apiError(response, payload);
    return payload as T;
  }

  async presigned(url: string, init: RequestInit): Promise<Response> {
    let parsed: URL;
    try {
      parsed = new URL(url);
    } catch {
      throw new JustDeployError('JustDeploy returned an invalid file URL.');
    }
    if (parsed.protocol !== 'https:' || parsed.username || parsed.password || parsed.hash) {
      throw new JustDeployError('JustDeploy returned an invalid file URL.');
    }

    try {
      return await this.fetcher(parsed, { ...init, redirect: 'error' });
    } catch {
      throw new JustDeployError(init.signal?.aborted ? 'The file transfer was canceled.' : 'The file transfer failed before the server returned a response.');
    }
  }

}

File: python/src/justdeploy/_transport.py
from __future__ import annotations

import json
from collections.abc import Mapping
from typing import Any, Literal, cast
from urllib.parse import quote, urlsplit

import httpx

from ._auth import SDK_HEADER, AsyncAuthManager, AuthSession, SyncAuthManager
from .errors import JustDeployError, JustDeployValidationError
from .types import AsyncUploadBody, SyncUploadBody

Method = Literal["GET", "POST", "PUT", "DELETE"]
API_TIMEOUT_SECONDS = 30.0
TRANSFER_FORBIDDEN_HEADERS = ("authorization", "x-justdeploy-sdk", "cookie")


def _api_error(response: httpx.Response, payload: object) -> JustDeployError:
    details = payload if isinstance(payload, dict) else {}
    message = details.get("message")
    if not isinstance(message, str) or not message:
        message = f"JustDeploy request failed with status {response.status_code}."
    retry_after = details.get("retryAfter")
    request_id = details.get("requestId")
    if not isinstance(request_id, str):
        request_id = response.headers.get("x-request-id")
    return JustDeployError(
        message,
        status=response.status_code,
        retry_after=retry_after if isinstance(retry_after, int) and not isinstance(retry_after, bool) else None,
        request_id=request_id,
        details=cast(Mapping[str, Any], details),
    )


def _payload(response: httpx.Response) -> object:
    if not response.content:
        return None
    try:
        return response.json()
    except (ValueError, UnicodeDecodeError):
        raise JustDeployError("JustDeploy returned a response that was not valid JSON.", status=response.status_code) from None


def _json_content(value: object) -> bytes:
    try:
        return json.dumps(value, ensure_ascii=False, allow_nan=False, separators=(",", ":")).encode()
    except (TypeError, ValueError, UnicodeEncodeError):
        raise JustDeployValidationError("The request contains a value that cannot be encoded as JSON.") from None


def _validate_presigned_url(url: str) -> str:
    try:
        parsed = urlsplit(url)
        _ = parsed.port
    except (TypeError, ValueError):
        raise JustDeployError("JustDeploy returned an invalid file URL.") from None
    if parsed.scheme != "https" or not parsed.hostname or parsed.username is not None or parsed.password is not None or parsed.fragment:
        raise JustDeployError("JustDeploy returned an invalid file URL.")
    return url


def _api_url(session: AuthSession, path: str) -> str:
    if not path.startswith("/") or path.startswith("//") or "://" in path or "\\" in path:
        raise RuntimeError("Invalid internal JustDeploy API path.")
    organization_id = quote(session.organization_id, safe="")
    return f"{session.api_origin}/organizations/{organization_id}{path}"


class SyncTransport:
    def __init__(self, client: httpx.Client, auth: SyncAuthManager) -> None:
        self.client = client
        self.auth = auth

    def organization_request(
        self,
        method: Method,
        path: str,
        *,
        json_body: object = None,
        headers: Mapping[str, str] | None = None,
    ) -> object:
        session = self.auth.get_session()
        return self._send(method, path, session, json_body=json_body, headers=headers, replayed=False)

    def _send(
        self,
        method: Method,
        path: str,
        session: AuthSession,
        *,
        json_body: object,
        headers: Mapping[str, str] | None,
        replayed: bool,
    ) -> object:
        request_headers = {"accept": "application/json", "authorization": f"Bearer {session.token}", "x-justdeploy-sdk": SDK_HEADER}
        for name, value in (headers or {}).items():
            if name.lower() in {"authorization", "host", "x-justdeploy-sdk"}:
                raise RuntimeError(f"The internal header {name} cannot be overridden.")
            request_headers[name.lower()] = value
        content = _json_content(json_body) if json_body is not None else None
        if content is not None:
            request_headers["content-type"] = "application/json"
        try:
            url = _api_url(session, path)
            response = self.client.request(
                method,
                url,
                headers=request_headers,
                content=content,
                timeout=API_TIMEOUT_SECONDS,
                follow_redirects=False,
            )
        except httpx.HTTPError:
            raise JustDeployError("The JustDeploy request failed before the server returned a response.") from None
        if response.status_code == 401 and method == "GET" and not replayed:
            response.close()
            refreshed = self.auth.refresh_after_unauthorized(session.token)
            return self._send(method, path, refreshed, json_body=json_body, headers=headers, replayed=True)
        payload = _payload(response)
        if not response.is_success:
            raise _api_error(response, payload)
        return payload

    def presigned_upload(self, url: str, *, mime: str, data: SyncUploadBody, size: int) -> httpx.Response:
        validated_url = _validate_presigned_url(url)
        try:
            request = self.client.build_request(
                "PUT",
                validated_url,
                headers={"content-type": mime, "content-length": str(size)},
                content=data,
            )
            for name in TRANSFER_FORBIDDEN_HEADERS:
                request.headers.pop(name, None)
            return self.client.send(request, follow_redirects=False)
        except Exception:
            raise JustDeployError("The file transfer failed before the server returned a response.") from None

    def presigned_download(self, url: str) -> httpx.Response:
        validated_url = _validate_presigned_url(url)
        try:
            request = self.client.build_request("GET", validated_url)
            for name in TRANSFER_FORBIDDEN_HEADERS:
                request.headers.pop(name, None)
            return self.client.send(request, stream=True, follow_redirects=False)
        except Exception:
            raise JustDeployError("The file transfer failed before the server returned a response.") from None


class AsyncTransport:
    def __init__(self, client: httpx.AsyncClient, auth: AsyncAuthManager) -> None:
        self.client = client
        self.auth = auth

    async def organization_request(
        self,
        method: Method,
        path: str,
        *,
        json_body: object = None,
        headers: Mapping[str, str] | None = None,
    ) -> object:
        session = await self.auth.get_session()
        return await self._send(method, path, session, json_body=json_body, headers=headers, replayed=False)

    async def _send(
        self,
        method: Method,
        path: str,
        session: AuthSession,
        *,
        json_body: object,
        headers: Mapping[str, str] | None,
        replayed: bool,
    ) -> object:
        request_headers = {"accept": "application/json", "authorization": f"Bearer {session.token}", "x-justdeploy-sdk": SDK_HEADER}
        for name, value in (headers or {}).items():
            if name.lower() in {"authorization", "host", "x-justdeploy-sdk"}:
                raise RuntimeError(f"The internal header {name} cannot be overridden.")
            request_headers[name.lower()] = value
        content = _json_content(json_body) if json_body is not None else None
        if content is not None:
            request_headers["content-type"] = "application/json"
        try:
            url = _api_url(session, path)
            response = await self.client.request(
                method,
                url,
                headers=request_headers,
                content=content,
                timeout=API_TIMEOUT_SECONDS,
                follow_redirects=False,
            )
        except httpx.HTTPError:
            raise JustDeployError("The JustDeploy request failed before the server returned a response.") from None
        if response.status_code == 401 and method == "GET" and not replayed:
            await response.aclose()
            refreshed = await self.auth.refresh_after_unauthorized(session.token)
            return await self._send(method, path, refreshed, json_body=json_body, headers=headers, replayed=True)
        payload = _payload(response)
        if not response.is_success:
            raise _api_error(response, payload)
        return payload

    async def presigned_upload(self, url: str, *, mime: str, data: AsyncUploadBody, size: int) -> httpx.Response:
        validated_url = _validate_presigned_url(url)
        try:
            request = self.client.build_request(
                "PUT",
                validated_url,
                headers={"content-type": mime, "content-length": str(size)},
                content=data,
            )
            for name in TRANSFER_FORBIDDEN_HEADERS:
                request.headers.pop(name, None)
            return await self.client.send(request, follow_redirects=False)
        except Exception:
            raise JustDeployError("The file transfer failed before the server returned a response.") from None

    async def presigned_download(self, url: str) -> httpx.Response:
        validated_url = _validate_presigned_url(url)
        try:
            request = self.client.build_request("GET", validated_url)
            for name in TRANSFER_FORBIDDEN_HEADERS:
                request.headers.pop(name, None)
            return await self.client.send(request, stream=True, follow_redirects=False)
        except Exception:
            raise JustDeployError("The file transfer failed before the server returned a response.") from None
